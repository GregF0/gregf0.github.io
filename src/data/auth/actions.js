import { firebase } from '../../firebase/firebase'

export const setUserInfo = (userInfo) => ({
  type: 'SET_USER_INFO',
  userInfo,
})

export const linkAccount = (authProvider) => {
  return async (dispatch) => {
    try {
      const res = await firebase.auth().currentUser.linkWithPopup(authProvider)

      const user = res.user

      const userInfo = { uid: user.uid }

      userInfo.creationTime = user.metadata.creationTime

      if (user.providerData && user.providerData.length) {
        userInfo.name = user.providerData[0].displayName
        userInfo.photo = user.providerData[0].photoURL
      }

      dispatch(setUserInfo(userInfo))
      return res
    } catch (e) {
      if (e.code === 'auth/credential-already-in-use') {
        return firebase.auth().signInWithCredential(e.credential)
      }
    }
  }
}

export const signOut = () => ({
  type: 'SIGN_OUT',
})

export const startSignOut = () => {
  return async (dispatch) => {
    await firebase.auth().signOut()
    dispatch(signOut())
  }
}
