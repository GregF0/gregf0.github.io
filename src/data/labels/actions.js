import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const newLabelRef = fs.collection(`users/${uid}/labels`).doc()

    dispatch(
      addLabel({
        id: newLabelRef.id,
        ...label,
      })
    )
    await newLabelRef.set({ ...label })
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(editLabel(id, updates))

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(deleteLabel(id))

    await fs.doc(`users/${uid}/labels/${id}`).delete()
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const labelsRef = await fs.collection(`users/${uid}/labels`).get()

      const labels = labelsRef.docs.map((label) => ({
        id: label.id,
        ...label.data(),
      }))

      dispatch(setLabels(labels))

      return labelsRef
    } catch (e) {
      dispatch(setLabels([]))
    }
  }
}
