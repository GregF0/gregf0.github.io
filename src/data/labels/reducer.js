export const initialState = {
  fullscreenDialog: false,
  desktopDialog: false,
  deleteAlert: false,
  menuOpened: false,
  formValue: null,
  labelEditting: null,
  labelSelected: null,
  data: [],
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_FULLSCREEN_DIALOG':
      return {
        ...state,
        fullscreenDialog: action.fullscreenDialog,
      }
    case 'SET_DESKTOP_DIALOG':
      return {
        ...state,
        desktopDialog: action.desktopDialog,
      }
    case 'SET_DELETE_ALERT':
      return {
        ...state,
        deleteAlert: action.deleteAlert,
      }
    case 'SET_MENU_OPENED':
      return {
        ...state,
        menuOpened: action.menuOpened,
      }
    case 'SET_LABEL_EDITTING':
      return {
        ...state,
        labelEditting: action.labelEditting,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    case 'ADD_LABEL':
      return {
        ...state,
        data: [...state.data, action.label],
        labelSelected: state.labelSelected || action.label,
      }
    case 'EDIT_LABEL':
      const updatedLabels = state.data.map((label) => {
        if (label.id === action.id) {
          return {
            ...label,
            ...action.updates,
          }
        } else {
          return label
        }
      })
      return {
        ...state,
        data: updatedLabels,
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? { ...state.labelSelected, ...action.updates }
            : state.labelSelected,
      }
    case 'DELETE_LABEL':
      return {
        ...state,
        data: state.data.filter(({ id }) => id !== action.id),
        labelSelected:
          state.labelSelected && state.labelSelected.id === action.id
            ? null
            : state.labelSelected,
      }
    case 'SET_LABELS':
      return {
        ...state,
        data: action.labels,
      }
    case 'SET_LABEL_SELECTED':
      return {
        ...state,
        labelSelected: action.labelSelected,
      }
    default:
      return state
  }
}
