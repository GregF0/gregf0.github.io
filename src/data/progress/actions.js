export const setProgressVisibility = (visibility) => ({
  type: 'SET_PROGRESS_VISIBILITY',
  visibility,
})
