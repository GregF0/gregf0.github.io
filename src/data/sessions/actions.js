import fs from '../../firebase/firebase'

export const setSessions = (sessions) => ({
  type: 'SET_SESSIONS',
  sessions,
})

export const startSetSessions = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const sessionsRef = await fs.collection(`users/${uid}/sessions`).get()

      const sessions = sessionsRef.docs.map((session) => ({
        id: session.id,
        ...session.data(),
      }))

      dispatch(setSessions(sessions))
      return sessionsRef
    } catch (e) {
      dispatch(setSessions([]))
    }
  }
}

export const addSession = (session) => ({
  type: 'ADD_SESSION',
  session,
})

export const startAddSession = (session) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    const { duration, label = null, createdAt } = session

    const newSessionRef = fs.collection(`users/${uid}/sessions`).doc()

    dispatch(
      addSession({
        id: newSessionRef.id,
        ...session,
      })
    )

    await newSessionRef.set({
      duration,
      label,
      createdAt,
    })

    return newSessionRef
  }
}
