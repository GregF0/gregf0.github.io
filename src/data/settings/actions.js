import fs from '../../firebase/firebase'

export const setWorkDuration = (duration) => ({
  type: 'SET_WORK_DURATION',
  duration,
})

export const startSetWorkDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setWorkDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { workDuration: duration } }, { merge: true })
  }
}

export const setShortBreakDuration = (duration) => ({
  type: 'SET_SHORT_BREAK_DURATION',
  duration,
})

export const startSetShortBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShortBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { shortBreakDuration: duration } }, { merge: true })
  }
}

export const setLongBreakDuration = (duration) => ({
  type: 'SET_LONG_BREAK_DURATION',
  duration,
})

export const startSetLongBreakDuration = (duration) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setLongBreakDuration(duration))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { longBreakDuration: duration } }, { merge: true })
  }
}

export const setRounds = (rounds) => ({
  type: 'SET_ROUNDS',
  rounds,
})

export const startSetRounds = (rounds) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setRounds(rounds))

    await fs.doc(`users/${uid}`).set({ settings: { rounds } }, { merge: true })
  }
}

export const setShowTimerInTitle = (showTimerInTitle) => ({
  type: 'SET_SHOW_TIMER_IN_TITLE',
  showTimerInTitle,
})

export const startSetShowTimerInTitle = (showTimerInTitle) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowTimerInTitle(showTimerInTitle))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showTimerInTitle } }, { merge: true })
  }
}

export const setShowNotifications = (showNotifications) => ({
  type: 'SET_SHOW_NOTIFICATIONS',
  showNotifications,
})

export const startSetShowNotifications = (showNotifications) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setShowNotifications(showNotifications))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { showNotifications } }, { merge: true })
  }
}

export const setDarkMode = (darkMode) => ({
  type: 'SET_DARK_MODE',
  darkMode,
})

export const startSetDarkMode = (darkMode) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    localStorage.setItem('darkMode', +darkMode)

    dispatch(setDarkMode(darkMode))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { darkMode } }, { merge: true })
  }
}

export const setAutostart = (autostart) => ({
  type: 'SET_AUTOSTART',
  autostart,
})

export const startSetAutostart = (autostart) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setAutostart(autostart))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { autostart } }, { merge: true })
  }
}

export const setFirstDayOfTheWeek = (firstDayOfTheWeek) => ({
  type: 'SET_FIRST_DAY_OF_THE_WEEK',
  firstDayOfTheWeek,
})

export const startSetFirstDayOfTheWeek = (firstDayOfTheWeek) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    dispatch(setFirstDayOfTheWeek(firstDayOfTheWeek))

    await fs
      .doc(`users/${uid}`)
      .set({ settings: { firstDayOfTheWeek } }, { merge: true })
  }
}

export const setSettings = (settings) => ({
  type: 'SET_SETTINGS',
  settings,
})

export const startSetSettings = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    try {
      const docRef = await fs.doc(`users/${uid}`).get()
      let data = {}

      if (docRef.exists) {
        data = docRef.data().settings
      }

      dispatch(setSettings(data))
    } catch (e) {
      dispatch(setSettings({}))
    }
  }
}
