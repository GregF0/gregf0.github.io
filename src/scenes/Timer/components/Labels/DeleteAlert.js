import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import styled from 'styled-components'
import Dialog from '@material-ui/core/Dialog'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import Button from '@material-ui/core/Button'
import {
  setLabelEditting,
  setFormValue,
  setFullscreenDialog,
  setDeleteAlert,
  startDeleteLabel,
} from '../../../../data/labels/actions'

export const DeleteAlert = () => {
  const { deleteAlert, labelEditting } = useSelector((state) => state.labels)

  const dispatch = useDispatch()

  const handleDelete = () => {
    dispatch(startDeleteLabel(labelEditting.id))

    dispatch(setDeleteAlert(false))
    dispatch(setFullscreenDialog(false))
    dispatch(setLabelEditting(null))
    dispatch(setFormValue(null))
  }

  const closeAlert = () => {
    dispatch(setDeleteAlert(false))
  }

  return (
    <Dialog
      open={deleteAlert}
      onClose={closeAlert}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
      id="delete-alert"
    >
      <DialogTitle id="alert-dialog-title">
        Delete
        <LabelName color={labelEditting && labelEditting.color}>
          {' '}
          {labelEditting && labelEditting.name}
        </LabelName>{' '}
        label?
      </DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">
          This action will also delete all the data and statistics related to
          this label.
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={closeAlert}>Cancel</Button>
        <Button onClick={handleDelete} color="secondary">
          Delete
        </Button>
      </DialogActions>
    </Dialog>
  )
}

const LabelName = styled.span`
  color: ${({ color }) => color};
`
