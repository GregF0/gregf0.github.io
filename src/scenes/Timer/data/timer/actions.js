export const setStatus = (status) => ({
  type: 'SET_STATUS',
  status,
})

export const setProgress = (progress) => ({
  type: 'SET_PROGRESS',
  progress,
})

export const setTimeLeft = (timeLeft) => ({
  type: 'SET_TIME_LEFT',
  timeLeft,
})

export const saveInterval = (interval) => ({
  type: 'SAVE_INTERVAL',
  interval,
})

export const pauseTimer = () => ({
  type: 'PAUSE_TIMER',
})

export const resetTimer = ({ duration, showTimerInTitle }) => ({
  type: 'RESET_TIMER',
  duration,
  showTimerInTitle,
})

export const setNextTimer = (settings) => ({
  type: 'SET_NEXT_TIMER',
  settings,
})

export const setSaveSessionAlert = (saveSessionAlert) => ({
  type: 'SET_SAVE_SESSION_ALERT',
  saveSessionAlert,
})
