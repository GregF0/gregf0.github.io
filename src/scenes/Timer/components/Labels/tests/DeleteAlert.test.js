import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import Dialog from '@material-ui/core/Dialog'
import Button from '@material-ui/core/Button'
import { DeleteAlert } from '../DeleteAlert'
import labels from '../../../../../data/labels/tests/mock-data/labels'
import * as labelsActions from '../../../../../data/labels/actions'

describe('<DeleteAlert />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<DeleteAlert />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  const createStore = (deleteAlert = false, labelEditting = null) => {
    const store = {
      labels: {
        deleteAlert,
        labelEditting,
      },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <DeleteAlert /> correctly', () => {
    createStore(true)
    expect(createWrapper()).toMatchSnapshot()
  })

  test('alert should be closed initially', () => {
    createStore(false)
    expect(createWrapper().find(Dialog).prop('open')).toBe(false)
  })

  test('alert should be opened ', () => {
    createStore(true)
    expect(createWrapper().find(Dialog).prop('open')).toBe(true)
  })

  test('should close alert on close button click ', () => {
    const setDeleteAlertMocked = jest
      .spyOn(labelsActions, 'setDeleteAlert')
      .mockImplementation(() => jest.fn())

    createStore(true)

    const wrapper = createWrapper()
    wrapper.find(Button).at(0).simulate('click')

    expect(setDeleteAlertMocked).toHaveBeenCalledWith(false)
  })

  test('should call startDeleteLabel on delete button click ', () => {
    const labelEdditing = labels[0]

    const startDeleteLabelMocked = jest
      .spyOn(labelsActions, 'startDeleteLabel')
      .mockImplementation(() => jest.fn())

    createStore(true, labelEdditing)

    createWrapper().find(Button).at(1).simulate('click')
    expect(startDeleteLabelMocked).toHaveBeenCalledWith(labelEdditing.id)
  })
})
