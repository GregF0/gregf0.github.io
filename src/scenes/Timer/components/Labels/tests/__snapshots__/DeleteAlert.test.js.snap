// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteAlert /> should render <DeleteAlert /> correctly 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-describedby="alert-dialog-description"
  aria-labelledby="alert-dialog-title"
  id="delete-alert"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="alert-dialog-title"
  >
    Delete
    <styled.span
      color={null}
    >
       
    </styled.span>
     
    label?
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))
      id="alert-dialog-description"
    >
      This action will also delete all the data and statistics related to this label.
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="secondary"
      onClick={[Function]}
    >
      Delete
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
